### DISCUSSION TITLE
Bus::fake() Construct Error

### DISCUSSION THREAD

**Discussion by TroyDavidson** [laravel/framework#43452] [Q&A]
I have a job that I need to test.  I've added `Bus::fake()` to my test, but it's throwing an error.

I've scaled the test way down and am still getting the error.  Here is the test:

```<?php

namespace Tests\Feature\Jobs;

use Illuminate\Support\Facades\Bus;
use Tests\TestCase;

class ExampleTest extends TestCase
{
    public function test_blah()
    {
        Bus::fake();

        $this->assertTrue(true);
    }
}
```

And this is the error:
```
 • Tests\Feature\Jobs\ExampleTest > blah
   TypeError 

  Illuminate\Support\Testing\Fakes\BusFake::__construct(): Argument #1 ($dispatcher) must be of type Illuminate\Contracts\Bus\QueueingDispatcher, null given, called in /Users/tdavidson/entrata/entrata-billpay/vendor/laravel/framework/src/Illuminate/Support/Facades/Bus.php on line 49

  at vendor/laravel/framework/src/Illuminate/Support/Testing/Fakes/BusFake.php:66
     62▕      * @param  \Illuminate\Contracts\Bus\QueueingDispatcher  $dispatcher
     63▕      * @param  array|string  $jobsToFake
     64▕      * @return void
     65▕      */
  ➜  66▕     public function __construct(QueueingDispatcher $dispatcher, $jobsToFake = [])
     67▕     {
     68▕         $this->dispatcher = $dispatcher;
     69▕ 
     70▕         $this->jobsToFake = Arr::wrap($jobsToFake);

      +1 vendor frames 
  2   tests/Feature/Jobs/ExampleTest.php:13
      Illuminate\Support\Facades\Bus::fake()


  Tests:  1 failed
  Time:   0.36s
  ```

I added an `echo "    facade root: '" . static::getFacadeRoot() . "'   ";` inside the fake() function:
```
    public static function fake($jobsToFake = [])
    {

        echo "    facade root: '" . static::getFacadeRoot() . "'   ";

        static::swap($fake = new BusFake(static::getFacadeRoot(), $jobsToFake));

        return $fake;
    }
```

The getFacadeRoot doesn't return anything.  I'm assuming this is the problem.  I'm running the test with `php artisan test`.  

I'm sure I'm missing something, but can't figure out what.  Any help would be great.

**Comment by henzeb** [comment]
Do you have `Illuminate\Bus\BusServiceProvider::class` registered correctly in your `config/app.php`

what does `app(Illuminate\Contracts\Bus\Dispatcher)` return?

**Comment by wizzymore** [comment]
I can confirm the behaviour on a new laravel application.

Also, `app(Illuminate\Contracts\Bus\Dispatcher::class)` throws `Target class [Tests\Unit\Illuminate\Contracts\Bus\Dispatcher] does not exist.` ( Only inside tests )

**Reply by henzeb** [reply]
And you are using the TestCase provided by Laravel? 

In your case it might also have something to do with the path, it is adding Tests\Unit. Try `app(\Illuminate\Contracts\Bus\Dispatcher::class)`

I've just updated my current project to the latest version, no issues.

**Reply by wizzymore** [reply]
Nvm, i am just dumb.
I was writing the test in the Unit folder.. Of course it can't work as laravel is not even booted.

Everything works correctly as a Feature test

**Reply by henzeb** [reply]
We've all been there... You can also boot laravel in side the Unit folder, just extend your test class with the TestCase provided by Laravel.
